### PR TITLE
BUG/API: preserve non-nano in factorize/unique

### LINKED ISSUE #51961: BUG (2.0rc0):  groupby changes dtype unexpectedly from `timedelta64[s]` to timedelta64[ns]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame(
    {"date": pd.date_range("2023-01-01", periods=5, unit="s"), "values": 1}
)

# datetime64[s]
print(df["date"].dtype)

grouped = df.groupby("date", as_index=False).sum()
# datetime64[ns]
print(grouped["date"].dtype)

concatted = pd.concat([df, grouped])["date"]
# object
print(concatted.dtype)
```

### Issue Description

I stumbled upon this behavior when test-driving 2.0rc0 with my code base.

Some data frames now contain date columns of type `datetime64[s]`. Did not investigate yet why. The problem occurs when working with these columns. After a `groupby`, the type changes to `datetime64[ns]`.

When now concatting `datetime64[s]` to `datetime64[ns]` things get even worse and I have an `object` column.

### Expected Behavior

As I started with a datetime column, I'd expect the datetime column dtype not to change magically. If a later `pd.merge` wouldn't have complained, it could have been hard to track down what was going on.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 1a2e300170efc08cb509a0b4ff6248f8d55ae777
python           : 3.11.0.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.1.0
Version          : Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:54 PDT 2022; root:xnu-8792.41.9~2/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 2.0.0rc0
numpy            : 1.24.2
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 67.4.0
pip              : 23.0.1
Cython           : None
pytest           : 7.2.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.2
html5lib         : 1.1
pymysql          : None
psycopg2         : 2.9.5
jinja2           : 3.1.2
IPython          : 8.11.0
pandas_datareader: None
bs4              : 4.11.2
bottleneck       : 1.3.7
brotli           : None
fastparquet      : None
fsspec           : 2023.1.0
gcsfs            : 2023.1.0
matplotlib       : 3.7.0
numba            : None
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.1.0
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : 1.4.46
tables           : None
tabulate         : 0.9.0
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None
</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #51961 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

The ugly bit here is that it causes `to_datetime` to behave slightly differently with cache=True/False in some cases.  I know @mroeschke worked to avoid that in #17077, so it may be a deal-breaker.

But it fixes #51961 (though doesn't yet have a targeted test) so may be worth prioritizing (cc @MarcoGorelli)

(Update) also fixes Series[non-nano].factorize which currently raises AssertionError

### CODE REVIEW

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -1076,31 +1076,39 @@ def test_to_datetime_array_of_dt64s(self, cache, unit):
...
+
         tm.assert_index_equal(result, expected)
 
         # A list of datetimes where the last one is out of bounds
         dts_with_oob = dts + [np.datetime64("9999-01-01")]
 
-        msg = "Out of bounds nanosecond timestamp: 9999-01-01 00:00:00"
-        with pytest.raises(OutOfBoundsDatetime, match=msg):
-            to_datetime(dts_with_oob, errors="raise")
+        # As of GH#?? we do not raise in this case
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:1090]: https://ichard26.github.io/next-pr-number/ 😉

File: pandas/tests/io/parser/test_parse_dates.py [hunk 2]
```diff
@@ -1657,8 +1657,8 @@ def date_parser(dt, time):
     datetimes = np.array(["2013-11-03T19:00:00"] * 3, dtype="datetime64[s]")
     expected = DataFrame(
         data={"rxstatus": ["00E80000"] * 3},
-        index=MultiIndex.from_tuples(
-            [(datetimes[0], 126), (datetimes[1], 23), (datetimes[2], 13)],
+        index=MultiIndex.from_arrays(
+            [datetimes, [126, 23, 13]],
```
Comment (MarcoGorelli) [inline pandas/tests/io/parser/test_parse_dates.py:1661]: why does this need to change?

**Review by MarcoGorelli — COMMENTED** [review]
Looks good to be honest, I don't think I'd consider the cache vs non-cache discrepancy a blocker

File: pandas/tests/io/parser/test_parse_dates.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/io/parser/test_parse_dates.py:1661]: with from_tuples it reconstructs a M8[ns] level.  with from_arrays it keeps the reso `datetimes` already has

File: pandas/tests/tools/test_to_datetime.py [hunk 3]
```diff
@@ -1076,31 +1076,39 @@ def test_to_datetime_array_of_dt64s(self, cache, unit):
         # Assuming all datetimes are in bounds, to_datetime() returns
         # an array that is equal to Timestamp() parsing
         result = to_datetime(dts, cache=cache)
-        expected = DatetimeIndex([Timestamp(x).asm8 for x in dts], dtype="M8[ns]")
+        if cache:
+            # FIXME: behavior should not depend on cache
+            expected = DatetimeIndex([Timestamp(x).asm8 for x in dts], dtype="M8[s]")
+        else:
+            expected = DatetimeIndex([Timestamp(x).asm8 for x in dts], dtype="M8[ns]")
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py:1083]: Do you have a inking what code path is casting the result to ns?

Would be good to have an issue for this once it's merged

Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1083]: array_to_datetime/maybe_convert_objects.  there's a checkbox for each in #46587

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1090]: have added this to my next CLN branch

**Comment by mroeschke** [comment]
Mind creating the backport PR @jbrockmendel ?

**Pull request merged**